### PR TITLE
meta program float-lib

### PR DESCRIPTION
**Pull request by cdonovick**
Probably won't pass `tests/test_constructors.py` until a new magma release is out

### CODE REVIEW

File: peak/float.py [hunk 1]
```diff
@@ -52,373 +55,222 @@ def to_rm(rm: BitVector[3]) -> RoundingMode:
...
+    @lru_cache(None)
+    def gen_binary(op_name):
+        @family_closure
+        def fc(family: TypeFamily):
+            cast = _get_cast
+            @family.assemble(locals(), globals())
+            class Op(Peak, BlackBox):
+                def __call__(self, rm: RoundingMode, in0: Data, in1: Data) -> Data:
+                    if rm == RoundingMode.RNE:
+                        in0_rne = cast(RM.RNE, in0)
```
Comment (cdonovick) [inline peak/float.py:98]: Basically of this is to cast between hwtypes rounding mode and floatlib rounding mode.

File: peak/float.py [hunk 2]
```diff
@@ -52,373 +55,222 @@ def to_rm(rm: BitVector[3]) -> RoundingMode:
...
     @family_closure
-    def Eq_fc(family: TypeFamily):
+    def fp_fma(family):
+        cast = _get_cast
         @family.assemble(locals(), globals())
-        class Eq(Peak, BlackBox):
-            def __call__(self, in0: Data, in1: Data) -> Data:
-                ...
-        return Eq
+        class FMA(Peak, BlackBox):
```
Comment (cdonovick) [inline peak/float.py:206]: special case FMA

File: peak/float.py [hunk 3]
```diff
@@ -52,373 +55,222 @@ def to_rm(rm: BitVector[3]) -> RoundingMode:
 
     return SimpleNamespace(**locals())
 
+
+def _format_name(name: str) -> str:
```
Comment (cdonovick) [inline peak/float.py:59]: To not break any code that might be using this already  I make sure the names haven't change

File: peak/float.py [hunk 4]
```diff
@@ -52,373 +55,222 @@ def to_rm(rm: BitVector[3]) -> RoundingMode:
...
+    assert name[:3] == 'fp_', name
+    return name[3:].capitalize() + '_fc'
+
 @lru_cache(None)
-def float_lib_gen(exp_bits: int, frac_bits: int):
+def float_lib_gen(exp_bits: int, frac_bits: int, ieee_compliance: bool = False):
```
Comment (cdonovick) [inline peak/float.py:64]: riscv is ieee compliant.

File: peak/float.py [hunk 5]
```diff
@@ -52,373 +55,222 @@ def to_rm(rm: BitVector[3]) -> RoundingMode:
...
 
-    @family_closure
-    def Gt_fc(family: TypeFamily):
-        @family.assemble(locals(), globals())
-        class Gt(Peak, BlackBox):
-            def __call__(self, in0: Data, in1: Data) -> Data:
-                ...
-        return Gt
+    closures = {}
+    for k, f in AbstractFPVector.__dict__.items():
```
Comment (cdonovick) [inline peak/float.py:195]: Generate a BB for each method in AbstractFPVector

**Review by rdaly525 — APPROVED** [review]
Looks good if all the tests pass

**Pull request merged**